### PR TITLE
[SPARK-23407][SQL] add a config to try to inline all mutable states during codegen

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/19811 .

In #19811, we picked a sub-optimal solution that always compact non-primitive mutable states to arrays, to make primitive mutable states more likely to get inlined.

This PR introduces a new config to not treat primitive states specially and try to inline all states, to avoid any potential perf regression in Spark 2.3. By default it's false.

In the future, we can remove this config, and dynamically decide which states to inline. For example, we can use placeholders during codegen, and analysis all the mutable states at the end and replace the placeholders.

Note that there are no known regression cases, so this is not a blocker for Spark 2.3

## How was this patch tested?

a new test.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @kiszk @rednaxelafx @viirya @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -675,6 +675,16 @@ object SQLConf {
       "disable logging or -1 to apply no limit.")
     .createWithDefault(1000)
 
+  val CODEGEN_TRY_INLINE_ALL_STATES =
+    buildConf("spark.sql.codegen.tryInlineAllStates")
+    .internal()
+    .doc("When adding mutable states during code generation, whether or not we should try to " +
+      "inline all the states. If this config is false, we only try to inline primitive stats, " +
+      "so that primitive states are more likely to be inlined. Set this config to true to make " +
+      "the behavior same as Spark 2.2.")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think it only behaves the same before we hit the threshold?

**Comment by mgaido91** [comment]
do we have any test about the performance regression introduced by the change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -1461,20 +1465,19 @@ object CodeGenerator extends Logging {
       CodegenMetrics.METRIC_GENERATED_CLASS_BYTECODE_SIZE.update(classBytes.length)
       try {
         val cf = new ClassFile(new ByteArrayInputStream(classBytes))
-        val stats = cf.methodInfos.asScala.flatMap { method =>
+        cf.methodInfos.asScala.flatMap { method =>
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Are these changes related to this PR?

**Comment by gatorsmile** [comment]
This PR will be hold until 2.3 is released.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: yea, let me improve it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: not, but a small clean up.

**Comment by cloud-fan** [comment]
@mgaido91 As I said in the PR description, no regression is found so far, just providing a config to be super safe.

Actually this PR has a problem: the codegen usually happens at executor side, so we can't use SQLConf directy. I'll figure this out after my vacation.

**Review by rednaxelafx — COMMENTED** [review]
You're right that in some cases codegen happens on the executors, so we can't use SQLConf directly.
In the case of whole-stage codegen, codegen happens on the driver side so that part is okay...

**Comment by kiszk** [comment]
retest this please
